### ISSUE TITLE
Overwrite allocating functions (e.g. `exp!`) for Circle?

### ISSUE THREAD

**Issue by kellertuer** [JuliaManifolds/Manifolds.jl#591]
Currently we have that the last line in 

```julia
using Manifolds
M = Circle()
p = 0.
X = π/2
q = exp(M,p,X)
exp!(M, q, p, X)
```

fails since that function does not exist for Circle (or `Euclidean()` or other number-type manifolds, I have a few listed at
https://github.com/JuliaManifolds/Manopt.jl/blob/9fc85ccf89f3aacbddcb57e3144fbe69fa23f62f/src/Manopt.jl#L193
This makes generic algorithms for them complicated, indeed quite a while ago I started 
https://github.com/JuliaManifolds/Manopt.jl/blob/master/src/plans/nonmutating_manifolds_plans.jl
which is only loaded if you load manifolds together with Manopt.jl

But that is code duplication (besides using `retract` in all places instead of `retract!`).

What do you think about introducing `retract!`? Is that possible or would that break stuff? We can not modify `q` sure, but can we find some other trick?

**Comment by mateuszbaran** [comment]
There is no way to make the last line in your example work without modifying previous lines. If you want to work on Circle with mutable functions, you need to use mutable objects to represent points, for example `fill(0.0)` or `[0.0]`.

**Comment by kellertuer** [comment]
Well, my main problem is maybe on the more generic side:

1) I want to write algorithms as efficient as pssible (hence use `retract!`) 
2) I wan to write them as generic as possible

currently 1 and 2 can not be achieved at the same time because either I am efficient (but not generic enough for circle) or I am generic (enough for circle, using `retract` instead) but not efficient.
I was wondering whether we can find a way to achieve both.

**Comment by mateuszbaran** [comment]
Mutable `retract!` works on all manifolds, just not for all point representations. I wouldn't worry too much about non-mutable representations in generic algorithms.

**Comment by kellertuer** [comment]
Well, as you can see, for the circle I had to do a full reimplementation of Armijo-linesearch to work and I fear that would hold for most algorithms. but I do not want to reimplement all of Manopt in that file to be able to work on the circle. That‘s where this question comes from.

**Comment by mateuszbaran** [comment]
If you do
```julia
using Manifolds
M = Circle()
p = [0.]
X = [π/2]
q = exp(M,p,X)
exp!(M, q, p, X)
```
it works -- you can use `Circle` with mutable functions. We can either use mutable representations or re-implement algorithms for non-mutable ones. I'm not sure what's your problem with mutable representation for `Circle`?

**Comment by kellertuer** [comment]
Well, my original data is more like data on the circle `[0., 0.1, 0.3]` so a whole array (ok maybe that could also be seen as data on power of circle). So maybe another power would also be nice here?

I just noticed I would like to get rid of reimplementing stuff, but if someone comes along with circle data, they have to modify their data.

**Comment by kellertuer** [comment]
Ah my test where I started that file was this mean, which is even on the circle.

https://github.com/JuliaManifolds/Manopt.jl/blob/9fc85ccf89f3aacbddcb57e3144fbe69fa23f62f/test/solvers/test_gradient_descent.jl#L4

**Comment by kellertuer** [comment]
Well,
I will think about this a bit. I plan to remove the duplicate code in Manopt 0.5, but I would prefer if the test case would still be able to run (I fear I might not find a way how).

**Comment by mateuszbaran** [comment]
Probably the easiest way would be handling immutability at the outer level of interface. Something like
```julia
function gradient_descent(M::AbstractManifold, f, grad_f, p::Number)
    q = gradient_descent(M, f, grad_f, fill(p))
    return q[]
end
function gradient_descent(M::AbstractManifold, f, grad_f, p)
    # generic implementation that assumes `p` is mutable
end
```

**Comment by kellertuer** [comment]
That is still quite a bit of boiler-plate code for the about 20 solvers we have, but looks nicer than the current approach for sure.

**Comment by mateuszbaran** [comment]
Yes, that's not ideal but I can't think of a better way of handling this. At least the boilerplate code is simple and not repeated algorithm logic.

**Comment by kellertuer** [comment]
Yes, not relating logic is much better than the current approach. I will think about that, it looks like 3-4 lines per solver, which is ok, and we have the two interfaces (with and without !) anyways which is a bit of boilerplate already, so maybe adding that would not hurt too much. Thanks for the idea (Maybe we should move this to Manopt then)

**Comment by mateuszbaran** [comment]
Yes, sure, I think this needs to be resolved in Manopt so the issue can be moved there. This should hopefully be easy to implement and non-breaking.
